### PR TITLE
Print a warning when single ZK replica with ephemeral storage is used

### PR DESCRIPTION
**Pull request by stanlyDoge**
Signed-off-by: Stanislav Knot <sknot@redhat.com>

### Type of change
- Enhancement 

### Description
While debugging faulty STs @tombentley noticed, we are using single node ZK clusters with ephemeral storage. It is ok, until we need any rolling update/restart. We do not want to disable this combination though. User should be at least log:WARNed the things will hit the fan eventually.

Fixes https://github.com/strimzi/strimzi-kafka-operator/issues/2555
### Checklist
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ZookeeperCluster.java [hunk 1]
```diff
@@ -181,6 +181,11 @@ public static ZookeeperCluster fromCrd(Kafka kafkaAssembly, KafkaVersion.Lookup
         if (replicas <= 0) {
             replicas = ZookeeperClusterSpec.DEFAULT_REPLICAS;
         }
+
+        if (replicas == 1 && zookeeperClusterSpec.getStorage() != null && zookeeperClusterSpec.getStorage().getType().equals("ephemeral")) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ZookeeperCluster.java]: Isn't `"ephemeral".equals(...)` better? Smaller chance for NPE.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ZookeeperCluster.java [hunk 2]
```diff
@@ -181,6 +181,11 @@ public static ZookeeperCluster fromCrd(Kafka kafkaAssembly, KafkaVersion.Lookup
...
+            log.warn("Zookeeper cluster with single replica and ephemeral storage will be in defective state after any restart or rolling update. It is recommended to use at least three replicas.");
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ZookeeperCluster.java]: @laidan6000 @PaulRMellor Could one of you review the message please?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ZookeeperCluster.java [hunk 3]
```diff
@@ -181,6 +181,11 @@ public static ZookeeperCluster fromCrd(Kafka kafkaAssembly, KafkaVersion.Lookup
         if (replicas <= 0) {
             replicas = ZookeeperClusterSpec.DEFAULT_REPLICAS;
         }
+
+        if (replicas == 1 && zookeeperClusterSpec.getStorage() != null && "ephemeral".equals(zookeeperClusterSpec.getStorage().getType())) {
+            log.warn("Zookeeper cluster with single replica and ephemeral storage will be in defective state after any restart or rolling update. It is recommended to use at least three replicas.");
```
Comment (PaulRMellor) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ZookeeperCluster.java]:
```suggestion
            log.warn("A Zookeeper cluster with a single replica and ephemeral storage will be in a defective state after any restart or rolling update. It is recommended that a minimum of three replicas are used.");
```

**Comment by dalelane** [comment]
@stanlyDoge I realize this is closed so I'm thinking of this a little late :-) 

But what do you think to also using the status for the Kafka custom resource to capture warnings like this, for users that don't have access to the cluster operator logs?

**Comment by tombentley** [comment]
@dalelane I'm not exactly opposed to it, but the operator doesn't currently have any machinery for collecting a list of extra conditions to be added when the status gets set. While it would be easy enough to hack something together, I think we really need an internal API in the operators which would set the status and issue the log warning together in a consistent, rather than ad-hoc way. Right now I think we just log warnings all in several places. Maybe fitting out `Reconciliation` with an append-only list of conditions would work, for example. 

It's also worth pointing out that the warning would be emitted immediately, whereas the status is usually only set at the end of a reconciliation. That mismatch could be confusing, for example, if the operator took a long time to finish the reconciliation due to pod rolling, or maybe got stuck because of the issue the warning was about (which would be a bug, but still delay the user diagnosing the problem). So ideally we'd have a mechanism to set the extra conditions or status during the reconciliation (and ignore the watch events which doing that would generate).

**Comment by scholzj** [comment]
I think it is easy to add the condition. You would just need to move the check from `KafkaCluster` class to `KafkaAssemnlyOperator`. Which should not be so hard if needed since all it does is to check replicas.

**Comment by dalelane** [comment]
The point about not having a one-off status set in a one-off place feels like a valid one. 

Taking a step back, what do you think about having a new issue about checking Kafka CRs for best practice, and logging warnings and updating status with advice/recommendations for anything found?

Ephemeral storage for single ZK replicas could be the first check on the list, but we could use the issue to come up with ideas for other helpful checks that we could do in the same way and place.

**Pull request merged**